Type stubs for tqdm library

Gilthans opened the pull request:
While tqdm is very popular, it is not only not typed, but strips the types of sequences it receives due to the way it works.
This should resolve the issue for most users.

An issue to add type hints to tqdm has been around [since 2016](https://github.com/tqdm/tqdm/issues/260), but hasn't seen any progress.

AlexWaygood commented:
It's fine to use `Literal` in typeshed, you just have to import it from `typing_extensions` instead of `typing`

AlexWaygood reviewed (commented):
Thanks for the PR! We'll take a full look when the tests are passing (feel free to ask if you need a hand with any of them).

Please don't make unrelated changes to `CONTRIBUTING.md` in this PR, though. If you think that needs to be changed, please do that in its own PR. It'll make for a very confusing git history otherwise.

AlexWaygood commented on stubs/tqdm/tqdm/dask.pyi:
```diff
@@ -0,0 +1,10 @@
+from _typeshed import Incomplete
+
+from dask.callbacks import Callback  # type: ignore
```
It's not currently possible to have third-party `py.typed` libraries as dependencies for typeshed stubs. Your best bet is to do something like this:

```python
from typing import Any
from typing_extensions import TypeAlias
Callback: TypeAlias = Any # actually dark.callbacks.Callback
```

Gilthans replied:
Thanks, I was wondering how to get around that!

AlexWaygood commented:
If the implementation breaks LSP by having an incompatible override, the best thing to do is to mimic the implementation exactly and add a `# type: ignore[override]` to silence mypy on the line in question

Gilthans commented:
Thanks for the feedback and tips, I've incorporated them!
There were failing tests in stubtest wherever there was an external dependency, so I added them to `stubtest_allowlist.txt`, I hope that's alright.

AlexWaygood commented:
> Thanks for the feedback and tips, I've incorporated them!
> There were failing tests in stubtest wherever there was an external dependency, so I added them to `stubtest_allowlist.txt`, I hope that's alright.

It looks like stubtest just needs to `pip install` a few more things before running the script. I think a better solution than allowlisting these modules is to add a `stubtest_requirements.txt` file to tell stubtest to install a few extra things. See here for an example: https://github.com/python/typeshed/blob/master/stubs/google-cloud-ndb/%40tests/requirements-stubtest.txt

AlexWaygood commented:
(You don't need to pin the requirements precisely -- the `google-cloud-ndb` stubs are a bit of an outlier in that regard!)

Gilthans commented:
I had to put `tqdm.contrib.discord` back to the allowlist, since the dependency has conflicts with tensorflow. Since it's a failry small, read-only [repo](https://github.com/b1naryth1ef/disco) I'm OK with not putting more effort into it, but if you have a suggestion I'll be happy to try something else

AlexWaygood commented:
> I had to put `tqdm.contrib.discord` back to the allowlist, since the dependency has conflicts with tensorflow. Since it's a failry small, read-only [repo](https://github.com/b1naryth1ef/disco) I'm OK with not putting more effort into it, but if you have a suggestion I'll be happy to try something else

Nope, that's totally cool -- thanks!

AlexWaygood commented on stubs/tqdm/tqdm/std.pyi:
```diff
@@ -0,0 +1,272 @@
+        bar_format: str | None = ...,
+        initial: float = ...,
+        position: int | None = ...,
+        postfix: Mapping[object, object] | str | None = ...,
```
Another `Mapping[object, object]`

AlexWaygood commented on stubs/tqdm/tqdm/std.pyi:
```diff
@@ -0,0 +1,272 @@
+    def __new__(cls, *_, **__): ...
+    @classmethod
+    def write(
+        cls, s: str, file: str | io.TextIOWrapper | io.StringIO | None = ..., end: str = ..., nolock: bool = ...
```
Again, the only required thing of the `file` parameter here is that the object passed in has a `write()` method:

```suggestion
        cls, s: str, file: SupportsWrite[str] | None = ..., end: str = ..., nolock: bool = ...
```

https://github.com/tqdm/tqdm/blob/4f208e72552c4d916aa4fe6a955349ee8b2ed353/tqdm/std.py#L613-L620

AlexWaygood commented on stubs/tqdm/tqdm/std.pyi:
```diff
@@ -0,0 +1,272 @@
+    def write(
+        cls, s: str, file: str | io.TextIOWrapper | io.StringIO | None = ..., end: str = ..., nolock: bool = ...
+    ) -> None: ...
+    @classmethod
+    def external_write_mode(
+        cls, file: str | io.TextIOWrapper | io.StringIO | None = ..., nolock: bool = ...
+    ) -> AbstractContextManager[None]: ...
```
`file` here should probably also be `SupportsWrite[str]`. Also, we prefer precise return types in typeshed, so it would be better to return `contextlib._GeneratorContextManager[None]` rather than `AbstractContextManager[None]`.

AlexWaygood commented on stubs/tqdm/tqdm/std.pyi:
```diff
@@ -0,0 +1,272 @@
+    last_print_t: Incomplete
+    start_t: Incomplete
+
+    def __bool__(self): ...
+    def __nonzero__(self): ...
+    def __len__(self): ...
+    def __reversed__(self): ...
+    def __contains__(self, item) -> bool: ...
+    def __enter__(self): ...
+    def __exit__(self, exc_type, exc_value, traceback) -> None: ...
```
```suggestion
    def __bool__(self) -> bool: ...
    def __nonzero__(self) -> bool: ...
    def __len__(self) -> int: ...
    def __reversed__(self) -> Iterator[_T]: ...
    def __contains__(self, item: object) -> bool: ...
    def __enter__(self: Self) -> Self: ...
    def __exit__(self, exc_type: object, exc_value: object, traceback: object) -> None: ...
```

AlexWaygood commented on stubs/tqdm/tqdm/std.pyi:
```diff
@@ -0,0 +1,272 @@
+    def reset(self, total: float | None = ...) -> None: ...
+    def set_description(self, desc: str | None = ..., refresh: bool | None = ...) -> None: ...
+    def set_description_str(self, desc: str | None = ..., refresh: bool | None = ...) -> None: ...
+    def set_postfix(self, ordered_dict: Mapping[object, object] | None = ..., refresh: bool | None = ..., **kwargs) -> None: ...
```
Another `Mapping[object, object]`

AlexWaygood commented on stubs/tqdm/tqdm/std.pyi:
```diff
@@ -0,0 +1,272 @@
+    def set_postfix_str(self, s: str = ..., refresh: bool = ...) -> None: ...
+    def moveto(self, n) -> None: ...
+    @property
+    def format_dict(self): ...
```
```suggestion
    def format_dict(self) -> MutableMapping[str, Any]: ...
```

https://github.com/tqdm/tqdm/blob/4f208e72552c4d916aa4fe6a955349ee8b2ed353/tqdm/std.py#L1460-L1476

AlexWaygood commented on stubs/tqdm/tqdm/std.pyi:
```diff
@@ -0,0 +1,272 @@
+    bar_format: str | None = ...,
+    initial: float = ...,
+    position: int | None = ...,
+    postfix: Mapping[object, object] | str | None = ...,
```
Another `Mapping[object, object]`

AlexWaygood commented on stubs/tqdm/tqdm/std.pyi:
```diff
@@ -0,0 +1,272 @@
+    bar_format: str | None = ...,
+    initial: float = ...,
+    position: int | None = ...,
+    postfix: Mapping[object, object] | str | None = ...,
```
Another `Mapping[object, object]`

AlexWaygood commented on stubs/tqdm/tqdm/utils.pyi:
```diff
@@ -0,0 +1,44 @@
+from _typeshed import Incomplete
+
+CUR_OS: Incomplete
+IS_WIN: Incomplete
+IS_NIX: Incomplete
+RE_ANSI: Incomplete
```
```suggestion
from _typeshed import Incomplete
from typing import Pattern

CUR_OS: str
IS_WIN: bool
IS_NIX: bool
RE_ANSI: Pattern[str]
```

AlexWaygood commented on stubs/tqdm/tqdm/utils.pyi:
```diff
@@ -0,0 +1,44 @@
+RE_ANSI: Incomplete
+
+class FormatReplace:
+    replace: Incomplete
+    format_called: int
+    def __init__(self, replace: str = ...) -> None: ...
+    def __format__(self, _): ...
```
```suggestion
    replace: str
    format_called: int
    def __init__(self, replace: str = ...) -> None: ...
    def __format__(self, _) -> str: ...
```

AlexWaygood commented on stubs/tqdm/tqdm/utils.pyi:
```diff
@@ -0,0 +1,44 @@
+    def __format__(self, _): ...
+
+class Comparable:
+    def __lt__(self, other): ...
+    def __le__(self, other): ...
+    def __eq__(self, other): ...
+    def __ne__(self, other): ...
+    def __gt__(self, other): ...
+    def __ge__(self, other): ...
```
```suggestion
    def __lt__(self, other) -> bool: ...
    def __le__(self, other) -> bool: ...
    def __eq__(self, other) -> bool: ...
    def __ne__(self, other) -> bool: ...
    def __gt__(self, other) -> bool: ...
    def __ge__(self, other) -> bool: ...
```

AlexWaygood commented on stubs/tqdm/tqdm/utils.pyi:
```diff
@@ -0,0 +1,44 @@
+    def __ge__(self, other): ...
+
+class ObjectWrapper:
+    def __getattr__(self, name): ...
+    def __setattr__(self, name, value): ...
```
```suggestion
    def __getattr__(self, name: str): ...
    def __setattr__(self, name: str, value): ...
```

AlexWaygood commented on stubs/tqdm/tqdm/utils.pyi:
```diff
@@ -0,0 +1,44 @@
+
+class SimpleTextIOWrapper(ObjectWrapper):
+    def __init__(self, wrapped, encoding) -> None: ...
+    def write(self, s): ...
+    def __eq__(self, other): ...
```
```suggestion
    def write(self, s: str): ...
    def __eq__(self, other: object) -> bool: ...
```

AlexWaygood commented on stubs/tqdm/tqdm/utils.pyi:
```diff
@@ -0,0 +1,44 @@
+    def __init__(self, wrapped, encoding) -> None: ...
+    def write(self, s): ...
+    def __eq__(self, other): ...
+
+class DisableOnWriteError(ObjectWrapper):
+    @staticmethod
+    def disable_on_exception(tqdm_instance, func): ...
```
```suggestion

_P = ParamSpec("_P")
_R = TypeVar("_R")

class DisableOnWriteError(ObjectWrapper):
    @staticmethod
    def disable_on_exception(tqdm_instance, func: Callable[_P, _R]) -> Callable[_P, _R]: ...
```

(You'll need to import `TypeVar` from `typing`, `ParamSpec` from `typing_extensions` and `Callable` from `collections.abc`)

AlexWaygood commented on stubs/tqdm/tqdm/utils.pyi:
```diff
@@ -0,0 +1,44 @@
+    @staticmethod
+    def disable_on_exception(tqdm_instance, func): ...
+    def __init__(self, wrapped, tqdm_instance) -> None: ...
+    def __eq__(self, other): ...
```
```suggestion
    def __eq__(self, other: object) -> bool: ...
```

AlexWaygood commented on stubs/tqdm/tqdm/utils.pyi:
```diff
@@ -0,0 +1,44 @@
+    def __eq__(self, other): ...
+
+class CallbackIOWrapper(ObjectWrapper):
+    def __init__(self, callback, stream, method: str = ...): ...
```
```suggestion
    def __init__(self, callback: Callable[[int], object], stream, method: str = ...) -> None: ...
```

AlexWaygood commented on stubs/tqdm/tqdm/utils.pyi:
```diff
@@ -0,0 +1,44 @@
+class CallbackIOWrapper(ObjectWrapper):
+    def __init__(self, callback, stream, method: str = ...): ...
+
+def disp_len(data): ...
+def disp_trim(data, length): ...
```
```suggestion
def disp_len(data: str) -> int: ...
def disp_trim(data: str, length: int) -> str: ...
```

AlexWaygood reviewed (changes requested):
This is a pretty big library to contribute, so thanks for all the work done here!

Here's a first pass at reviewing -- sorry about the number of comments! Most are pretty minor.

I'll do another pass after these have been addressed. I haven't looked at the `mypy_primer` report yet.

AlexWaygood commented on stubs/tqdm/tqdm/notebook.pyi:
```diff
@@ -0,0 +1,93 @@
+        bar_format: str | None = ...,
+        initial: float = ...,
+        position: int | None = ...,
+        postfix: Mapping[object, object] | str | None = ...,
```
This is marked as resolved but doesn't seem to be fixed

AlexWaygood replied:
```suggestion
        postfix: Mapping[str, object] | str | None = ...,
```

AlexWaygood commented on stubs/tqdm/tqdm/notebook.pyi:
```diff
@@ -0,0 +1,93 @@
+        bar_format: str | None = ...,
+        initial: float = ...,
+        position: int | None = ...,
+        postfix: Mapping[object, object] | str | None = ...,
```
Also unresolved

AlexWaygood replied:
```suggestion
        postfix: Mapping[str, object] | str | None = ...,
```

AlexWaygood commented on stubs/tqdm/tqdm/dask.pyi:
```diff
@@ -0,0 +1,12 @@
+_Callback: TypeAlias = Any  # Actually dask.callbacks.Callback
+
+class TqdmCallback(_Callback):
+    tqdm_class: Incomplete
+    def __init__(
+        self, start: Incomplete | None = ..., pretask: Incomplete | None = ..., tqdm_class=..., **tqdm_kwargs
```
```suggestion
    tqdm_class: type[Incomplete]
    def __init__(
        self, start: Incomplete | None = ..., pretask: Incomplete | None = ..., tqdm_class: type[Incomplete] = ..., **tqdm_kwargs
```

AlexWaygood commented on stubs/tqdm/tqdm/gui.pyi:
```diff
@@ -0,0 +1,28 @@
+    def clear(self, *_, **__) -> None: ...
+    def display(self, *_, **__) -> None: ...
+
+def tgrange(*args, **kwargs): ...
```
```suggestion
def tgrange(*args, **kwargs) -> tqdm_gui[Incomplete]: ...
```

AlexWaygood commented on stubs/tqdm/tqdm/keras.pyi:
```diff
@@ -0,0 +1,29 @@
+        verbose: int = ...,
+        tqdm_class=...,
+        **tqdm_kwargs,
+    ): ...
```
```suggestion
    ) -> None: ...
```

AlexWaygood commented on stubs/tqdm/tqdm/keras.pyi:
```diff
@@ -0,0 +1,29 @@
+        **tqdm_kwargs,
+    ): ...
+    def on_train_begin(self, *_, **__) -> None: ...
+    def on_epoch_begin(self, epoch, *_, **__): ...
```
```suggestion
    def on_epoch_begin(self, epoch, *_, **__) -> None: ...
```

AlexWaygood commented on stubs/tqdm/tqdm/rich.pyi:
```diff
@@ -0,0 +1,36 @@
+    def display(self, *_, **__) -> None: ...
+    def reset(self, total: Incomplete | None = ...) -> None: ...
+
+def trrange(*args, **kwargs): ...
```
```suggestion
def trrange(*args, **kwargs) -> tqdm_rich[Incomplete]: ...
```

AlexWaygood commented on stubs/tqdm/tqdm/std.pyi:
```diff
@@ -0,0 +1,270 @@
+        delay: float | None = ...,
+        gui: bool = ...,
+    ) -> None: ...
+    def __new__(cls, *_, **__): ...
```
```suggestion
    def __new__(cls: type[Self], *_, **__) -> Self: ...
```

AlexWaygood commented on stubs/tqdm/tqdm/std.pyi:
```diff
@@ -0,0 +1,270 @@
+        nrows: int | None = ...,
+        colour: str | None = ...,
+        delay: float | None = ...,
+    ): ...
```
```suggestion
    ) -> None: ...
```

AlexWaygood commented on stubs/tqdm/tqdm/std.pyi:
```diff
@@ -0,0 +1,270 @@
+    @classmethod
+    def wrapattr(
+        cls, stream, method: Literal["read", "write"], total: float | None = ..., bytes: bool | None = ..., **tqdm_kwargs
+    ) -> AbstractContextManager[Incomplete]: ...
```
```suggestion
    ) -> contextlib._GeneratorContextManager[Incomplete]: ...
```

AlexWaygood commented on stubs/tqdm/tqdm/tk.pyi:
```diff
@@ -0,0 +1,23 @@
+    def cancel(self) -> None: ...
+    def reset(self, total: Incomplete | None = ...) -> None: ...
+
+def ttkrange(*args, **kwargs): ...
```
```suggestion
def ttkrange(*args, **kwargs) -> tqdm_tk[Incomplete]: ...
```

AlexWaygood reviewed (changes requested):
Second pass. We're nearly there.

AlexWaygood commented on stubs/tqdm/tqdm/tk.pyi:
```diff
@@ -0,0 +1,23 @@
+_T = TypeVar("_T")
+
+class tqdm_tk(Generic[_T], std_tqdm[_T]):
+    def __init__(self, *args, **kwargs): ...
```
```suggestion
    def __init__(self, *args, **kwargs) -> None: ...
```

AlexWaygood commented on stubs/tqdm/tqdm/utils.pyi:
```diff
@@ -0,0 +1,49 @@
+
+class ObjectWrapper:
+    def __getattr__(self, name: str): ...
+    def __setattr__(self, name: str, value): ...
```
```suggestion
    def __setattr__(self, name: str, value) -> None: ...
```

AlexWaygood commented on stubs/tqdm/tqdm/contrib/utils_worker.pyi:
```diff
@@ -0,0 +1,15 @@
+
+class MonoWorker:
+    pool: ThreadPoolExecutor
+    futures: deque[Incomplete]
```
```suggestion
    futures: deque[Future[Incomplete]]
```

AlexWaygood commented:
(I might need to iterate a little bit to tackle these primer errors)

AlexWaygood reviewed (approved):
Awesome, thanks for your patience! The remaining mypy_primer hits all look like true positives to me (except the `pyppeteer` one, which I can't make head or tail of... but you can't catch them all `¯\_(ツ)_/¯`)

JelleZijlstra commented:
Weird, https://github.com/pyppeteer/pyppeteer/blob/f98f9030d3d40dc629bf08f93672f05caf8dfb0e/pyppeteer/chromium_downloader.py#L94 doesn't have a `tqdm` line. The next line does, but there's no `file` argument. Is mypy-primer looking at a different version of the library? The file hasn't changed since January.

AlexWaygood commented:
> Weird, https://github.com/pyppeteer/pyppeteer/blob/f98f9030d3d40dc629bf08f93672f05caf8dfb0e/pyppeteer/chromium_downloader.py#L94 doesn't have a `tqdm` line. The next line does, but there's no `file` argument. Is mypy-primer looking at a different version of the library? The file hasn't changed since January.

Yeah, I have no idea. Weirdly, the default branch for `pyppeteer` seems to be `dev` rather than `master` — but the error doesn't make sense on `master` either.

JelleZijlstra commented:
There's a hardcoded revision: https://github.com/hauntsaninja/mypy_primer/blob/d91de1f533963a41d362531147c2ff407c286433/mypy_primer.py#L1431

This is the version it's checking: https://github.com/pyppeteer/pyppeteer/blob/2d27bfdf9b6d0df32e5ebe869b057409042417a8/pyppeteer/chromium_downloader.py#L94
It's passing `os.devnull` which is indeed a `str`, not a `SupportsWrite[str]`. I don't know whether that code is just broken.

AlexWaygood commented:
> There's a hardcoded revision: https://github.com/hauntsaninja/mypy_primer/blob/d91de1f533963a41d362531147c2ff407c286433/mypy_primer.py#L1431
> 
> This is the version it's checking: https://github.com/pyppeteer/pyppeteer/blob/2d27bfdf9b6d0df32e5ebe869b057409042417a8/pyppeteer/chromium_downloader.py#L94 It's passing `os.devnull` which is indeed a `str`, not a `SupportsWrite[str]`. I don't know whether that code is just broken.

Hmm, seems like primer's been hardcoded to that revision for 2 years now. I wonder if `tqdm` has changed in the meantime so that `str` is no longer good to be passed to the `file` parameter. It doesn't look to me like `file` is meant to be a `str` (but I haven't tried it, just read the source code)

alkatar21 commented:
@AlexWaygood  I do not know if I should create a new issue or if it will be seen here. Anyway, under `tqdm.contrib.logging` are 2 context managers that are currently annotated as iterator and therefore generate the following messages:
```py
    with logging_redirect_tqdm():
        pass
```
```
"Iterator[None]" has no attribute "__enter__"; maybe "__iter__"?  [attr-defined] mypy(error)
"Iterator[None]" has no attribute "__exit__"; maybe "__next__"?  [attr-defined] mypy(error)
Object of type "Iterator[None]" cannot be used with "with" because it does not implement __enter__ Pylance[reportGeneralTypeIssues]
Object of type "Iterator[None]" cannot be used with "with" because it does not implement __exit__ Pylance[reportGeneralTypeIssues]
```

AlexWaygood commented:
@alkatar21 thanks! I've filed #8251.
